app/ModelFunctions/SymLinkFunctions.php

```diff
@@ -1,31 +1,53 @@
 <?php
 
-/** @noinspection PhpUndefinedClassInspection */
-
 namespace App\ModelFunctions;
 
 use App;
-use App\Photo;
 use App\Configs;
+use App\Photo;
 use App\SymLink;
 use Storage;
 
 class SymLinkFunctions
 {
+	/**
+	 * @var SessionFunctions
+	 */
+	private $sessionFunctions;
+
+	/**
+	 * AlbumFunctions constructor.
+	 *
+	 * @param SessionFunctions $sessionFunctions
+	 */
+	public function __construct(SessionFunctions $sessionFunctions)
+	{
+		$this->sessionFunctions = $sessionFunctions;
+	}
+
 	/**
 	 * @param Photo $photo
 	 *
 	 * @return SymLink|null
 	 */
 	public function find(Photo $photo)
 	{
-		$sym = null;
+		if (Storage::getDefaultDriver() == 's3') {
+			return null;
+		}
+		if (Configs::get_value('SL_enable', '0') === '0') {
+			return null;
+		}
 
-		if (Storage::getDefaultDriver() == 's3' || Configs::get_value('enable_picture_symlink', '0') === '0') {
-			return $sym;
+		if ($this->sessionFunctions->is_admin() && Configs::get_value('SL_for_admin', '0') === '0') {
+			return null;
 		}
 
-		$sym = SymLink::where('photo_id', $photo->id)->orderBy('created_at', 'DESC')->first();
+		$sym = null;
+
+		$sym = SymLink::where('photo_id', $photo->id)
+			->orderBy('created_at', 'DESC')
+			->first();
```

Comment:
I was meaning to ask: how can we have more than one SymLink row per photo? Is that due to race conditions?

Reply:
Race condition was not in my mind but more the fact that we may want to extend the work to 1 symlink per visitor... I don't know yet...

Reply:
At one per visitor per photo it *must* be slower than an auth check...

Reply:
@d7415 actually no even at one per visitor per photo it is not slower.
The reason is you only generate the symlink once. After that it is only a `get` on the symlink.
When the visitor access the picture we don't have check, it is directly served with the content of the symlink.
A authorization based system would require to serve the `image/size/id` url, go through Laravel for **each** image call before actually serving the image.